File: packages/domain/src/entities/span.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpanRecord {
    pub trace_id: String,
    pub span_id: String,
    pub parent_span_id: Option<String>,
    pub name: String,
    pub kind: i32,
    pub start_time_ns: u64,
    pub end_time_ns: u64,
    pub duration_ns: u64,
    pub status_code: i32,
    pub status_message: Option<String>,
    pub trace_state: Option<String>,
    pub flags: u32,
    pub dropped_attributes_count: u32,
    pub dropped_events_count: u32,
    pub dropped_links_count: u32,
    pub service_name: String,
    pub resource_attributes: Value,
    pub resource_dropped_attributes_count: u32,
    pub resource_schema_url: Option<String>,
    pub scope_name: Option<String>,
    pub scope_version: Option<String>,
    pub scope_attributes: Value,
    pub scope_dropped_attributes_count: u32,
    pub scope_schema_url: Option<String>,
    pub attributes: Value,
    pub events: Value,
    pub links: Value,
}

File: apps/api/src/features/traces/types/span.ts
import type { IngestProviderName, Json } from "@shared/helpers"

export type SpanRecord = {
  traceId: string
  spanId: string
  parentSpanId?: string
  name: string
  kind: number
  startTimeNs: bigint
  endTimeNs: bigint
  durationNs: bigint
  statusCode: number
  statusMessage?: string
  traceState?: string
  flags: number
  droppedAttributesCount: number
  droppedEventsCount: number
  droppedLinksCount: number
  serviceName: string
  resourceAttributes: Json
  resourceDroppedAttributesCount: number
  resourceSchemaUrl?: string
  scopeName?: string
  scopeVersion?: string
  scopeAttributes: Json
  scopeDroppedAttributesCount: number
  scopeSchemaUrl?: string
  attributes: Json
  events: Json
  links: Json
  ingestProvider?: IngestProviderName
}

export type TraceStatus = "ok" | "error"

export type TraceSummary = {
  traceId: string
  rootSpanId?: string
  rootObserved: boolean
  rootService?: string
  rootName?: string
  startTimeNs: bigint
  endTimeNs: bigint
  durationNs: bigint
  status: TraceStatus
  spanCount: number
  httpMethod?: string
  httpStatusCode?: number
  httpUrl?: string
  httpRoute?: string
  breakdown: BreakdownItem[] | null
}

export type BreakdownItem = {
  name: string
  durationNs: number
  spanCount: number
}

export type TraceSortField =
  | "date"
  | "root_service"
  | "name"
  | "duration"
  | "spans"
  | "status"

export type TraceSortOrder = "asc" | "desc"

export type TraceListFilters = {
  limit: number
  offset: number
  sort: TraceSortField
  order: TraceSortOrder
  service?: string
  status?: string
  method?: string
  httpStatusCode?: number
  name?: string
  url?: string
  durationMinNs?: bigint
  durationMaxNs?: bigint
  sinceNs?: bigint
}

export type FacetValue = {
  value: string
  count: number
}

export type TraceFacets = {
  services: FacetValue[]
  statuses: FacetValue[]
  methods: FacetValue[]
  httpStatusCodes: FacetValue[]
  routes: FacetValue[]
  durations: FacetValue[]
}
